### PR TITLE
feat: InputGroup component

### PR DESCRIPTION
**Pull request by cheton**
# InputGroup

`InputGroup` is used to group related add-ons and inputs.

### Default imports

```js
import InputGroup from '@trendmicro/react-styled-core/InputGroup';
import InputGroupAddon from '@trendmicro/react-styled-core/InputGroupAddon';
import InputGroupAppend from '@trendmicro/react-styled-core/InputGroupAppend';
import InputGroupPrepend from '@trendmicro/react-styled-core/InputGroupPrepend';
```

### Named imports

```js
import {
  InputGroup,
  InputGroupAddon,
  InputGroupAppend,
  InputGroupPrepend,
} from '@trendmicro/react-styled-core';
```

## Usage

Use `InputGroup` to place an `InputGroupAddon` or `Button` on either side of an `Input`. You may also place one on both sides of an `Input`.

image

```jsx
<Stack direction="column" spacing="4x">
  <InputGroup>
    <InputGroupPrepend>
      <InputGroupAddon variant="filled">@</InputGroupAddon>
    </InputGroupPrepend>
    <Input placeholder="Username" />
  </InputGroup>
  <InputGroup>
    <Input placeholder="Recipient's username" />
    <InputGroupAppend>
      <InputGroupAddon variant="filled">@example.com</InputGroupAddon>
    </InputGroupAppend>
  </InputGroup>
  <InputGroup>
    <InputGroupPrepend>
      <InputGroupAddon variant="filled">$</InputGroupAddon>
    </InputGroupPrepend>
    <Input placeholder="Recipient's username" />
    <InputGroupAppend>
      <InputGroupAddon variant="filled">.00</InputGroupAddon>
    </InputGroupAppend>
  </InputGroup>
</Stack>
```

### Sizes

Use the `size` prop to change the size of the input group. You can set the value to `sm`, `md`, or `lg`.

image

```jsx
<Stack direction="column" spacing="4x">
  <InputGroup size="sm">
    <InputGroupPrepend>
      <InputGroupAddon variant="filled">sm</InputGroupAddon>
    </InputGroupPrepend>
    <Input placeholder="Small size (24px)" />
  </InputGroup>
  <InputGroup size="md">
    <InputGroupPrepend>
      <InputGroupAddon variant="filled">md</InputGroupAddon>
    </InputGroupPrepend>
    <Input placeholder="Default size (32px)" />
  </InputGroup>
  <InputGroup size="lg">
    <InputGroupPrepend>
      <InputGroupAddon variant="filled">lg</InputGroupAddon>
    </InputGroupPrepend>
    <Input placeholder="Large size (40px)" />
  </InputGroup>
</Stack>
```

### Variants

The `InputGroup` component comes in 3 variants: `outline`, `filled`, and `unstyled`. Pass the `variant` prop and set it to either of these values.

#### `outline`

image

```jsx
<Stack direction="column" spacing="4x">
  <InputGroup variant="outline">
    <InputGroupPrepend>
      <InputGroupAddon>@</InputGroupAddon>
    </InputGroupPrepend>
    <Input placeholder="Username" />
  </InputGroup>
  <InputGroup variant="outline">
    <InputGroupPrepend>
      <InputGroupAddon>@</InputGroupAddon>
    </InputGroupPrepend>
    <Input variant="filled" placeholder="Username" />
  </InputGroup>
</Stack>
```

#### `filled`

image

```jsx
<Stack direction="column" spacing="4x">
  <InputGroup variant="filled">
    <InputGroupPrepend>
      <InputGroupAddon variant="filled">@</InputGroupAddon>
    </InputGroupPrepend>
    <Input variant="outline" placeholder="Username" />
  </InputGroup>
  <InputGroup variant="filled">
    <InputGroupPrepend>
      <InputGroupAddon>@</InputGroupAddon>
    </InputGroupPrepend>
    <Input placeholder="Username" />
  </InputGroup>
</Stack>
```

#### `unstyled`

For an unstyled input group that doesn't have border and padding spaces, it is not necessary to place an `InputGroupPrepend` or `InputGroupAppend` on either side of an `Input`.

image

```jsx
<Stack direction="column" spacing="4x">
  <InputGroup variant="unstyled">
    <InputGroupAddon pr="1x">@</InputGroupAddon>
    <Input placeholder="Username" />
  </InputGroup>
</Stack>
```

### Multiple add-ons

Multiple add-ons are supported and can be mixed with checkbox and radio input components.

image

```jsx
<Stack direction="column" spacing="4x">
  <InputGroup>
    <InputGroupPrepend>
      <InputGroupAddon variant="filled">$</InputGroupAddon>
    </InputGroupPrepend>
    <InputGroupPrepend>
      <InputGroupAddon variant="filled">0.00</InputGroupAddon>
    </InputGroupPrepend>
    <Input />
  </InputGroup>
  <InputGroup>
    <Input />
    <InputGroupAppend>
      <InputGroupAddon variant="filled">$</InputGroupAddon>
    </InputGroupAppend>
    <InputGroupAppend>
      <InputGroupAddon variant="filled">0.00</InputGroupAddon>
    </InputGroupAppend>
  </InputGroup>
</Stack>
```

### Multiple inputs

While multiple `<Input />`s are supported visually, validation styles are only available for input groups with a single `<Input />`.

image

```jsx
<InputGroup>
  <InputGroupPrepend>
    <InputGroupAddon variant="filled">First and last name</InputGroupAddon>
  </InputGroupPrepend>
  <Input placeholder="First name" defaultValue="John" required />
  <Input placeholder="Last name" defaultValue="Doe" required />
</InputGroup>
```

### Date inputs

image

```jsx
<InputGroup variant="unstyled">
  <Input
    placeholder="YYYY"
    borderBottom={1}
    textAlign="center"
    width="12x"
  />
  <Text px="2x">–</Text>
  <Input
    placeholder="MM"
    borderBottom={1}
    textAlign="center"
    width="8x"
  />
  <Text px="2x">–</Text>
  <Input
    placeholder="DD"
    borderBottom={1}
    textAlign="center"
    width="8x"
  />
</InputGroup>
```

### Button add-ons

image

```jsx
```jsx
<Stack direction="column" spacing="4x">
  <InputGroup>
    <InputGroupPrepend mr={0}>
      <Button>Button</Button>
    </InputGroupPrepend>
    <Input />
  </InputGroup>
  <InputGroup>
    <Input />
    <InputGroupAppend ml={0}>
      <Button>Button</Button>
    </InputGroupAppend>
  </InputGroup>
  <InputGroup>
    <InputGroupPrepend mr={0}>
      <Button variant="primary">Button</Button>
    </InputGroupPrepend>
    <InputGroupPrepend>
      <Button variant="secondary">Button</Button>
    </InputGroupPrepend>
    <Input />
  </InputGroup>
  <InputGroup>
    <Input />
    <InputGroupAppend ml={0}>
      <Button variant="primary">Button</Button>
    </InputGroupAppend>
    <InputGroupAppend>
      <Button variant="secondary">Button</Button>
    </InputGroupAppend>
  </InputGroup>
</Stack>
```

### Checkbox and radio add-ons

Place any checkbox or radio option within an input group's add-on instead of text.

image

```jsx
<Stack direction="column" spacing="4x">
  <InputGroup>
    <InputGroupPrepend>
      <InputGroupAddon variant="filled">
        <Checkbox />
      </InputGroupAddon>
    </InputGroupPrepend>
    <Input placeholder="Text input with checkbox" />
  </InputGroup>
  <InputGroup>
    <InputGroupPrepend>
      <InputGroupAddon variant="filled">
        <Radio />
      </InputGroupAddon>
    </InputGroupPrepend>
    <Input placeholder="Text input with radio button" />
  </InputGroup>
</Stack>
```

### InputGroup

`InputGroup` composes the [`Box`](./box) component. You can override the default styles with style props.

| Name | Type | Default | Description |
| :--- | :--- | :------ | :---------- |
| `size` | `string` | 'md' | The relative size to the input group itself. One of: 'sm', 'md', 'lg' |
| `variant` | `string` | 'outline' | The variant of the input style to use. One of: 'outline', 'filled', 'unstyled' |

### InputGroupAddon

`InputGroupAddon` composes the [`Flex`](./flex) component.

| Name | Type | Default | Description |
| :--- | :--- | :------ | :---------- |
| `size` | `string` | 'md' | The relative size to the input group itself. One of: 'sm', 'md', 'lg' |
| `variant` | `string` | 'outline' | The variant of the input style to use. One of: 'outline', 'filled', 'unstyled' |

### InputGroupAppend

`InputGroupAppend` composes the [`Flex`](./flex) component.

### InputGroupPrepend

`InputGroupPrepend` composes the [`Flex`](./flex) component.

### CODE REVIEW

File: packages/react-styled-core/src/InputGroup/index.js [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+            size: child.props.size ?? size,
+            variant: child.props.variant ?? variant,
+          };
+
+          if (child.type === Input) {
+            const inputProps = getInputProps(childProps);
+            return cloneElement(child, inputProps);
+          }
+
+          return cloneElement(child, childProps);
```
Comment (tinaClin) [inline packages/react-styled-core/src/InputGroup/index.js]: 感覺可以不用cloneElement，直接render children，利用context取得props

Comment (cheton) [reply packages/react-styled-core/src/InputGroup/index.js]: Resolved in https://github.com/trendmicro-frontend/styled-ui/pull/81/commits/5509b599657e8862c729394d89004100c18cc31b

File: packages/react-styled-core/src/Input/index.js [hunk 2]
```diff
@@ -1,17 +1,57 @@
...
+  if (inputGroupContext) {
+    const {
+      size: inputGroupSize,
+      variant: inputGroupVariant,
+    } = { ...inputGroupContext };
+
+    // - Use the inherited value from the input group
+    // - Fallback to the default value if the value is null or undefined
+    size = (size ?? inputGroupSize) ?? defaultSize;
+    variant = (variant ?? inputGroupVariant) ?? defaultSize;
```
Comment (tinaClin) [inline packages/react-styled-core/src/Input/index.js]: 要`defaultVariant `唷！

Comment (cheton) [reply packages/react-styled-core/src/Input/index.js]: 要!

Comment (cheton) [reply packages/react-styled-core/src/Input/index.js]: 如果沒給 size 也沒有 inputGroupSize 就要 fallback 到 defaultSize

Comment (cheton) [reply packages/react-styled-core/src/Input/index.js]: fixed in 9762e499706c68d4f59385623f8595b87dc38bac

Thank you @tinaClin

File: packages/react-styled-core/src/InputGroupAddon/index.js [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+const InputGroupAddon = forwardRef((
+  {
+    size,
+    variant,
+    ...rest
+  },
+  ref,
+) => {
+  const {
```
Comment (tinaClin) [inline packages/react-styled-core/src/InputGroupAddon/index.js:19]: `InputGroupAddon`在取context時沒有像`Input`有先判斷 `if (inputGroupContext)`，如果單獨使用`InputGroupAddon`時，應該會發生javascript error...這邊要加判斷嗎？還是？

File: packages/react-styled-core/src/InputGroupAddon/styles.js [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+  const color = {
+    dark: 'white:primary',
+    light: 'black:primary',
+  }[colorMode];
+
+  return {
+    color,
+    border: 0,
+    borderRadius: 0,
+    px: undefined,
```
Comment (tinaClin) [inline packages/react-styled-core/src/InputGroupAddon/styles.js:77]: `undefined`? `unset`?

File: packages/react-styled-core/src/InputGroupAddon/index.js (hunk 3 above)
Comment (cheton) [reply packages/react-styled-core/src/InputGroupAddon/index.js:19]: 本來是靠 throw error 丟球給 developer，在改掉 useInputGroup 的作法之後應該是要處理 exception，我來改一下

File: packages/react-styled-core/src/InputGroupAddon/styles.js (hunk 4 above)
Comment (cheton) [reply packages/react-styled-core/src/InputGroupAddon/styles.js:77]: yes! 要用 `undefined` 把 `px` 給取消掉，Chakra UI 也是這樣的做法

File: packages/react-styled-core/src/InputGroupAddon/index.js (hunk 3 above)
Comment (cheton) [reply packages/react-styled-core/src/InputGroupAddon/index.js:19]: fixed in 0d9250514cfc2b3a688de7fe5acf0b9a40e39004

**Review by tinaClin — APPROVED** [review]

**Pull request merged**